src/array/cuda/csr_sum.cu

```diff
@@ -0,0 +1,168 @@
+  const DType* B_weights = B_weights_array.Ptr<DType>();
+  // allocate cusparse handle if needed
+  if (!thr_entry->cusparse_handle)
+    CUSPARSE_CALL(cusparseCreate(&(thr_entry->cusparse_handle)));
+  CUSPARSE_CALL(cusparseSetStream(thr_entry->cusparse_handle, thr_entry->stream));
+
+  cusparseMatDescr_t matA, matB, matC;
+  CUSPARSE_CALL(cusparseCreateMatDescr(&matA));
+  CUSPARSE_CALL(cusparseCreateMatDescr(&matB));
+  CUSPARSE_CALL(cusparseCreateMatDescr(&matC));
+
+  cusparseSetPointerMode(thr_entry->cusparse_handle, CUSPARSE_POINTER_MODE_HOST);
+  size_t workspace_size = 0;
+  /* prepare output C */
+  IdArray dC_csrOffsets = IdArray::Empty({A.num_rows+1}, A.indptr->dtype, ctx);
+  IdType* dC_csrOffsets_data = dC_csrOffsets.Ptr<IdType>();
+  IdArray dC_columns;
+  NDArray dC_weights;
+  IdType* dC_columns_data = dC_columns.Ptr<IdType>();
+  DType* dC_weights_data = dC_weights.Ptr<DType>();
+  /* prepare buffer */
+  CUSPARSE_CALL(CSRGEAM<DType>::bufferSizeExt(
+      thr_entry->cusparse_handle, m, n, &alpha,
+      matA, nnzA, A_weights,
+      A.indptr.Ptr<IdType>(),
+      A.indices.Ptr<IdType>(),
+      &beta, matB, nnzB, B_weights,
+      B.indptr.Ptr<IdType>(),
+      B.indices.Ptr<IdType>(),
+      matC, dC_weights_data, dC_csrOffsets_data, dC_columns_data,
+      &workspace_size));
+
+  void *workspace = device->AllocWorkspace(ctx, workspace_size);
+  CUSPARSE_CALL(CSRGEAM<DType>::nnz(thr_entry->cusparse_handle,
+      m, n, matA, nnzA,
+      A.indptr.Ptr<IdType>(),
+      A.indices.Ptr<IdType>(),
+      matB, nnzB,
+      B.indptr.Ptr<IdType>(),
+      B.indices.Ptr<IdType>(),
+      matC, dC_csrOffsets_data, &nnzC, workspace));
+
+  dC_columns = IdArray::Empty({nnzC}, A.indptr->dtype, ctx);
+  dC_weights = NDArray::Empty({nnzC}, A_weights_array->dtype, ctx);
+  dC_columns_data = dC_columns.Ptr<IdType>();
+  dC_weights_data = dC_weights.Ptr<DType>();
+
+  CUSPARSE_CALL(CSRGEAM<DType>::compute(
+      thr_entry->cusparse_handle, m, n, &alpha,
+      matA, nnzA, A_weights,
+      A.indptr.Ptr<IdType>(),
+      A.indices.Ptr<IdType>(),
+      &beta, matB, nnzB, B_weights,
+      B.indptr.Ptr<IdType>(),
+      B.indices.Ptr<IdType>(),
+      matC, dC_weights_data, dC_csrOffsets_data, dC_columns_data,
+      workspace));
+
+  device->FreeWorkspace(ctx, workspace);
+  // destroy matrix/vector descriptors
+  CUSPARSE_CALL(cusparseDestroyMatDescr(matA));
+  CUSPARSE_CALL(cusparseDestroyMatDescr(matB));
+  CUSPARSE_CALL(cusparseDestroyMatDescr(matC));
+  return {CSRMatrix(A.num_rows, A.num_cols, dC_csrOffsets, dC_columns),
+    dC_weights};
+}
+}  // namespace cusparse
+
+template <int XPU, typename IdType, typename DType>
+std::pair<CSRMatrix, NDArray> CSRSum(
+    const std::vector<CSRMatrix>& As,
+    const std::vector<NDArray>& A_weights) {
+  CHECK(As.size() > 0) << "List of matrices can't be empty.";
+  CHECK_EQ(As.size(), A_weights.size()) << "List of matrices and weights must have same length";
```

Comment:
Consider moving these checks to device-agnostic function (e.g., in `src/array/kernel.cc`).